src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/EnumConverter.cs

```diff
@@ -43,7 +46,9 @@ public EnumConverter(EnumConverterOptions converterOptions, JsonNamingPolicy? na
         {
             _converterOptions = converterOptions;
             _namingPolicy = namingPolicy;
+            _serializerOptions = serializerOptions;
             _nameCache = new ConcurrentDictionary<ulong, JsonEncodedText>();
+            _sourceNameCache = new ConcurrentDictionary<JsonEncodedText, string>();
```

Comment:
It could be better to use the encoded string version of the enum value as `TKey` and the enum value as `TValue` so when you call `TryGetValue` while reading you don't have to create the JsonEncodedText which IIRC is expensive.
```suggestion
            _sourceNameCache = new ConcurrentDictionary<string, T>();
```

Reply:
Thanks for the code review.

1. I replace cache type to ConcurrentDictionary<string, string>
2. Initialize only need add new value.
3. And add support Flags
4. Add test case TestFlagsEnumValuesWithNamingPolicy

and update this pull request.

I just don't know what to do in cases where the NameCacheSizeSoftLimit = 64 restriction occurs.

Reply:
And I'll try to see how it is done in Newtonsoft.Json StringEnumConverter.

Reply:
On Newtonsoft.Json create cache by all enum names, maybe there is nothing dangerous when once (in constructor or static code) a cache is created for a type for all its values? 

Is there a danger of a cache overflow only if it is added to the cache at the user's request (for example, write json)?

Current version of type EnumConverter contains two point adding values to cache: 1) constructor and 2) method Write(...). I think that NameCacheSizeSoftLimit added when cache writes on method Write(...).

Maybe adding cache values in method's write & writequotes not actual? And check NameCacheSizeSoftLimit in constructor not actual?

https://github.com/dotnet/runtime/blob/b8a66770f205691745445890509c0f06537313a9/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/EnumConverter.cs#L388

https://github.com/dotnet/runtime/blob/b8a66770f205691745445890509c0f06537313a9/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/EnumConverter.cs#L199

Reply:
I realized that adding to the cache in Write(...) methods is required in the case of flags when intermediate values.
However, in the constructor, it may not be worth checking the NameCacheSizeSoftLimit?